crates/turborepo-lockfiles/src/berry/identifiers.rs

```diff
@@ -0,0 +1,371 @@
+            .as_ref()
+            .map(|scope| scope.to_string())
+            .map(Cow::Owned);
+        let name = Cow::Owned(name.to_string());
+        Ident { scope, name }
+    }
+}
+
+// These TryFrom impls should be FromStr, but to avoid unnecessary copying we
+// use TryFrom so we can use a lifetime.
+impl<'a> TryFrom<&'a str> for Ident<'a> {
+    type Error = Error;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        let make_err = || Error::Ident(value.to_string());
+        let captures = IDENT.captures(value).ok_or_else(make_err)?;
+        let scope = captures.get(1).map(|m| Cow::Borrowed(m.as_str()));
+        let name = Cow::Borrowed(captures.get(2).map(|m| m.as_str()).ok_or_else(make_err)?);
+        Ok(Self { scope, name })
+    }
+}
+
+impl<'a> fmt::Display for Ident<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(scope) = self.scope.as_deref() {
+            f.write_fmt(format_args!("@{scope}/"))?;
+        }
+        f.write_str(&self.name)
+    }
+}
+
+impl<'a> TryFrom<&'a str> for Descriptor<'a> {
+    type Error = Error;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        let make_err = || Error::Descriptor(value.to_string());
+        let captures = DESCRIPTOR.captures(value).ok_or_else(make_err)?;
+        let scope = captures.get(1).map(|m| Cow::Borrowed(m.as_str()));
+        let name = Cow::Borrowed(captures.get(2).map(|m| m.as_str()).ok_or_else(make_err)?);
+        let range = Cow::Borrowed(captures.get(3).map(|m| m.as_str()).ok_or_else(make_err)?);
+        let ident = Ident { scope, name };
+        Ok(Descriptor { ident, range })
+    }
+}
+
+impl<'a> fmt::Display for Descriptor<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_fmt(format_args!("{}@{}", self.ident, self.range))
+    }
+}
+
+impl<'a> Descriptor<'a> {
+    pub fn new(ident: &'a str, range: &'a str) -> Result<Self, Error> {
+        let ident = Ident::try_from(ident)?;
+        let range = range.into();
+        Ok(Self { ident, range })
+    }
+
+    /// Extracts all descriptors that are present in a lockfile entry key
+    pub fn from_lockfile_key(key: &'a str) -> impl Iterator<Item = Result<Descriptor<'a>, Error>> {
+        MULTIKEY.split(key).map(Descriptor::try_from)
+    }
+
+    /// Removes the protocol from a version range
+    pub fn strip_protocol(range: &str) -> &str {
+        range
+            .find(':')
+            .map_or(range, |colon_index| &range[colon_index + 1..])
+    }
+
+    pub fn into_owned(self) -> Descriptor<'static> {
+        let Self { ident, range } = self;
+        let range = Cow::Owned(range.to_string());
+        Descriptor {
+            ident: ident.to_owned(),
+            range,
+        }
+    }
+
+    /// Returns the protocol of the version range if one is present
+    pub fn protocol(&self) -> Option<&str> {
+        self.range.find(':').map(|i| &self.range[0..i])
```

Comment:
I quite like split_once for this since it avoids thinking about off-by-one

```suggestion
        self.range.split_once(':').map(|(p,_)| p)
```